File: rag/src/normalizer.py
"""
normalizer.py - Markdown 文档标准化器。

读取 Markdown 源文件，并根据路径和内容线索生成 RAG 索引需要的
document 元数据结构。
"""

import re
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent

SOURCE_TO_CATEGORY = {
    "human-linguistics": {
        "rules": "common.humanization",
        "references": "common.language",
    },
    "plot-review": {
        "rules": "common.outline_review",
        "references": "common.consistency",
    },
    "plot-ideation": {
        "rules": "common.ideation",
        "references": "common.ideation",
    },
    "rhythm-review": {
        "rules": "common.rhythm",
        "references": "common.rhythm",
    },
    "game-datafied": {
        "rules": "common.outline",
        "references": "common.outline",
    },
    "project-templates": {
        "project": "common.project",
    },
    "knowledge": {
        "deslop": "common.humanization",
        "topics": "common.outline",
        "writing": "common.ideation",
        "market": "common.project",
        "cases": "common.consistency",
    },
}

FILE_CATEGORY_OVERRIDES = {
    "章节写前准备清单": "common.prewrite",
    "大纲质量评估清单": "common.outline_review",
    "状态一致性补充清单": "common.consistency",
    "去AI味最小修改指南": "common.humanization",
    "句式节奏档案": "common.rhythm",
    "阅读体验与章节润色检查": "common.rhythm",
}


def _extract_skill_name(rel_path):
    parts = rel_path.replace("\\", "/").split("/")
    try:
        skills_idx = parts.index("skills")
        if skills_idx + 1 < len(parts):
            return parts[skills_idx + 1]
    except ValueError:
        pass
    return None


def _infer_stage(rel_path, content):
    stages = []
    lower = content.lower()
    path_lower = rel_path.lower()

    if any(k in path_lower for k in ("ideation", "灵感", "构思")):
        stages.append("ideation")
    if any(k in path_lower for k in ("outline", "大纲", "结构")):
        stages.append("outline")
    if any(k in path_lower for k in ("prewrite", "写前", "准备")):
        stages.append("prewrite")
    if any(k in path_lower for k in ("rhythm", "节奏", "润色", "阅读体验")):
        stages.extend(["drafting", "revision"])
    if any(k in path_lower for k in ("human", "去ai", "修改", "自然")):
        stages.append("revision")
    if any(k in path_lower for k in ("review", "审查", "评估", "审稿")):
        stages.append("review")

    if any(k in lower for k in ("大纲", "骨架", "结构评估")):
        if "outline" not in stages:
            stages.append("outline")
        if "review" not in stages:
            stages.append("review")
    if any(k in lower for k in ("写前", "准备", "开始写")):
        if "prewrite" not in stages:
            stages.append("prewrite")
    if any(k in lower for k in ("修改", "润色", "去ai", "总结腔")):
        if "revision" not in stages:
            stages.append("revision")

    if not stages:
        stages.extend(["drafting", "review"])

    return list(dict.fromkeys(stages))


def _infer_doc_id(rel_path):
    path = rel_path.replace("\\", "/")
    file_name = Path(rel_path).stem

    skill_name = _extract_skill_name(rel_path)
    if skill_name:
        return f"{skill_name}.{file_name}"

    if path.startswith(".harness/knowledge/"):
        parts = path.split("/")
        if len(parts) >= 5:
            return f"knowledge.{parts[3]}.{file_name}"
        return f"knowledge.{file_name}"

    if path.startswith(".harness/project-templates/"):
        return f"project.{file_name}"

    return file_name


def _infer_category(rel_path):
    path = rel_path.replace("\\", "/")
    file_name = Path(rel_path).stem

    if file_name in FILE_CATEGORY_OVERRIDES:
        return FILE_CATEGORY_OVERRIDES[file_name]

    if path.startswith(".harness/project-templates/"):
        return "common.project"

    if path.startswith(".harness/knowledge/"):
        parts = path.split("/")
        if len(parts) >= 5:
            section = parts[3]
            return SOURCE_TO_CATEGORY["knowledge"].get(section, "common")
        return "common"

    skill_name = _extract_skill_name(rel_path)
    if not skill_name:
        return "common"

    is_rule = "/rules/" in path
    is_ref = "/references/" in path

    mapping = SOURCE_TO_CATEGORY.get(skill_name)
    if mapping:
        if is_rule and "rules" in mapping:
            return mapping["rules"]
        if is_ref and "references" in mapping:
            return mapping["references"]

    return "common"


def _infer_scope(rel_path):
    path = rel_path.replace("\\", "/")
    if path.startswith(".harness/project-templates/"):
        return "project"
    return "common"


def _infer_tags(rel_path, content):
    tags = []
    path = rel_path.replace("\\", "/")
    file_name = Path(rel_path).stem

    tags.append(file_name)

    skill_name = _extract_skill_name(rel_path)
    if skill_name:
        tags.append(skill_name)

    if path.startswith(".harness/knowledge/"):
        parts = path.split("/")
        tags.extend([p for p in parts[2:-1] if p not in ("included", "remote")])

    title_match = re.search(r"^#\s+(.+)", content, re.MULTILINE)
    if title_match:
        tags.append(title_match.group(1).strip())

    return tags


def _infer_priority(rel_path):
    path = rel_path.replace("\\", "/")
    if "/rules/" in path:
        return 1
    if path.startswith(".harness/knowledge/"):
        return 2
    if "/references/" in path:
        return 2
    if path.startswith(".harness/project-templates/"):
        return 3
    return 4


def normalize_document(rel_path, source_type):
    """把单个源文件标准化为 document 字典。"""
    abs_path = PROJECT_ROOT / rel_path

    try:
        content = abs_path.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError) as exc:
        print(f"[normalizer] 无法读取文件: {rel_path} ({exc})")
        return None

    title_match = re.search(r"^#\s+(.+)", content, re.MULTILINE)
    title = title_match.group(1).strip() if title_match else Path(rel_path).stem

    return {
        "doc_id": _infer_doc_id(rel_path),
        "title": title,
        "source_path": rel_path.replace("\\", "/"),
        "source_type": source_type,
        "category": _infer_category(rel_path),
        "stage": _infer_stage(rel_path, content),
        "scope": _infer_scope(rel_path),
        "tags": _infer_tags(rel_path, content),
        "priority": _infer_priority(rel_path),
        "status": "active",
        "_raw_content": content,
    }

File: rag/src/normalizer.js
/**
 * normalizer.js — 文档标准化器
 *
 * 职责：
 * - 读取 Markdown 文件原始内容
 * - 从路径和内容中提取 metadata
 * - 输出统一的 document 结构
 */

import { readFileSync } from 'fs';
import { join, resolve, dirname, basename, extname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = resolve(__dirname, '../../');

/**
 * source_type 到 category 的映射规则
 */
const SOURCE_TO_CATEGORY = {
  'human-linguistics': {
    rules: 'common.humanization',
    references: 'common.language'
  },
  'plot-review': {
    rules: 'common.outline_review',
    references: 'common.consistency'
  },
  'plot-ideation': {
    rules: 'common.ideation',
    references: 'common.ideation'
  },
  // 特殊文件名映射
  FILE_CATEGORY_OVERRIDES: {
    '章节写前准备清单': 'common.prewrite',
    '大纲质量评估清单': 'common.outline_review',
    '状态一致性补充清单': 'common.consistency',
    '去AI味最小修改指南': 'common.humanization',
    '句式节奏档案': 'common.rhythm',
    '阅读体验与章节润色检查': 'common.rhythm',
  },
  'rhythm-review': {
    rules: 'common.rhythm',
    references: 'common.rhythm'
  },
  'game-datafied': {
    rules: 'common.outline',
    references: 'common.outline'
  },
  'projects': {
    project: 'common.project'
  }
};

/**
 * 从相对路径中提取技能目录名
 */
function extractSkillName(relPath) {
  const parts = relPath.replace(/\\/g, '/').split('/');
  // .harness/skills/{skill_name}/...
  const skillsIdx = parts.indexOf('skills');
  if (skillsIdx >= 0 && skillsIdx + 1 < parts.length) {
    return parts[skillsIdx + 1];
  }
  return null;
}

/**
 * 确定 stage（适用阶段）
 */
function inferStage(relPath, content) {
  const stages = [];
  const lower = content.toLowerCase();
  const pathLower = relPath.toLowerCase();

  // 从路径推断
  if (pathLower.includes('ideation') || pathLower.includes('灵感') || pathLower.includes('构思')) {
    stages.push('ideation');
  }
  if (pathLower.includes('outline') || pathLower.includes('大纲') || pathLower.includes('结构')) {
    stages.push('outline');
  }
  if (pathLower.includes('prewrite') || pathLower.includes('写前') || pathLower.includes('准备')) {
    stages.push('prewrite');
  }
  if (pathLower.includes('rhythm') || pathLower.includes('节奏') || pathLower.includes('润色') || pathLower.includes('阅读体验')) {
    stages.push('drafting', 'revision');
  }
  if (pathLower.includes('human') || pathLower.includes('去ai') || pathLower.includes('修改') || pathLower.includes('自然')) {
    stages.push('revision');
  }
  if (pathLower.includes('review') || pathLower.includes('审查') || pathLower.includes('评估') || pathLower.includes('审稿')) {
    stages.push('review');
  }

  // 从内容推断
  if (lower.includes('大纲') || lower.includes('骨架') || lower.includes('结构评估')) {
    if (!stages.includes('outline')) stages.push('outline');
    if (!stages.includes('review')) stages.push('review');
  }
  if (lower.includes('写前') || lower.includes('准备') || lower.includes('开始写')) {
    if (!stages.includes('prewrite')) stages.push('prewrite');
  }
  if (lower.includes('修改') || lower.includes('润色') || lower.includes('去ai') || lower.includes('总结腔')) {
    if (!stages.includes('revision')) stages.push('revision');
  }

  // 默认
  if (stages.length === 0) {
    stages.push('drafting', 'review');
  }

  return [...new Set(stages)];
}

/**
 * 从路径推断 doc_id
 */
function inferDocId(relPath) {
  const parts = relPath.replace(/\\/g, '/').split('/');
  const fileName = basename(parts[parts.length - 1], '.md');

  const skillName = extractSkillName(relPath);
  if (skillName) {
    return `${skillName}.${fileName}`;
  }

  // projects
  if (relPath.startsWith('.harness/projects/')) {
    return `project.${fileName}`;
  }

  return fileName;
}

/**
 * 从路径推断 category
 */
function inferCategory(relPath) {
  const path = relPath.replace(/\\/g, '/');
  const fileName = basename(relPath, '.md');

  // 文件名覆盖（处理特殊文档映射）
  const overrides = SOURCE_TO_CATEGORY.FILE_CATEGORY_OVERRIDES || {};
  if (overrides[fileName]) {
    return overrides[fileName];
  }

  // projects
  if (path.startsWith('.harness/projects/')) {
    return 'common.project';
  }

  const skillName = extractSkillName(relPath);
  if (!skillName) return 'common';

  const isRule = path.includes('/rules/');
  const isRef = path.includes('/references/');

  const mapping = SOURCE_TO_CATEGORY[skillName];
  if (mapping) {
    if (isRule && mapping.rules) return mapping.rules;
    if (isRef && mapping.references) return mapping.references;
  }

  return 'common';
}

/**
 * 推断 scope
 */
function inferScope(relPath, priority = 3) {
  if (relPath.startsWith('.harness/projects/')) {
    return 'project';
  }
  return 'common';
}

/**
 * 提取标签
 */
function inferTags(relPath, content) {
  const tags = [];
  const path = relPath.replace(/\\/g, '/');
  const fileName = basename(relPath, '.md');

  // 从文件名提取标签
  tags.push(fileName);

  // 从路径提取
  const skillName = extractSkillName(relPath);
  if (skillName) tags.push(skillName);

  // 从内容标题提取
  const titleMatch = content.match(/^#\s+(.+)/m);
  if (titleMatch) {
    const title = titleMatch[1].trim();
    tags.push(title);
  }

  return tags;
}

/**
 * 推断优先级 (1=最高, 5=最低)
 */
function inferPriority(relPath) {
  const path = relPath.replace(/\\/g, '/');
  // rule > reference > project
  if (path.includes('/rules/')) return 1;
  if (path.includes('/references/')) return 2;
  if (path.startsWith('.harness/projects/')) return 3;
  return 4;
}

/**
 * 标准化单个文件为 document 对象
 * @param {string} relPath - 相对路径
 * @param {string} sourceType - 源类型 (rule/reference/project)
 * @returns {object} document 对象
 */
export function normalizeDocument(relPath, sourceType) {
  const absPath = join(PROJECT_ROOT, relPath);

  let content = '';
  try {
    content = readFileSync(absPath, 'utf-8');
  } catch (e) {
    console.warn(`[normalizer] 无法读取文件: ${relPath}`, e.message);
    return null;
  }

  // 提取标题
  const titleMatch = content.match(/^#\s+(.+)/m);
  const title = titleMatch ? titleMatch[1].trim() : basename(relPath, '.md');

  const docId = inferDocId(relPath);
  const category = inferCategory(relPath);
  const priority = inferPriority(relPath);
  const tags = inferTags(relPath, content);
  const stage = inferStage(relPath, content);

  return {
    doc_id: docId,
    title,
    source_path: relPath.replace(/\\/g, '/'),
    source_type: sourceType,
    category,
    stage,
    scope: inferScope(relPath, priority),
    tags,
    priority,
    status: 'active',
    _raw_content: content  // 内部传递，不持久化
  };
}

export default { normalizeDocument };
